[BE] Improve Flight Recorder efficacy

c-p-i-o opened the pull request:
Summary:
This is an attempt to improve the flight recorder efficacy. 
We have a small subset of jobs that are timing out (i.e. failing to write out FR logs in 1 minute) and some that are throwing a `std::exception - broken promise`.

There are two changes in here.
1. We attempt to write out FR buffer with stack traces. If this fails, we attempt to capture FR buffer again - but this time without stack traces. The assumption here is that FR could be locking up when unwinding stack.
Note, to keep things simple, I'm re-using the same file name for both with/without stack_trace.
2.  Add additional catch statements in the Manifold writer. There might be something going on in here - so we'll get a log statement if this is failing.


TODO:
- there's nothing differentiating in the output that says whether stack traces were omitted purposefully or not.
This info might be useful for the analyzer - so I'll add this in a follow on diff.

Test Plan: Unit tests.

Differential Revision: D66843194




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1729,16 +1735,30 @@ void ProcessGroupNCCL::heartbeatMonitor() {
   if (checkDumpSignal && shouldDump_.load()) {
     // Store debug info to storage if no other thread does it. (By default to
     // local disk)
-    std::future<bool> asyncDebugDump = std::async(
+    std::future<bool> asyncDebugDumpWithStack = std::async(
         std::launch::async, [this]() { return this->dumpDebuggingInfo(); });
 
     // wait for the dump until timeout - log data
-    waitForFutureOrTimeout(
-        asyncDebugDump,
+    auto result = waitForFutureOrTimeout(
+        asyncDebugDumpWithStack,
         std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
         "Flight recorder dump in heartbeatMonitor",
         false,
         true);
+
+    // If we failed to dump, try dumping without stack trace.
+    if (!result) {
+      std::future<bool> asyncDebugDumpWithNoStack =
+          std::async(std::launch::async, [this]() {
+            return this->dumpDebuggingInfo(/*captureStackTrace=*/false);
+          });
+      waitForFutureOrTimeout(
+          asyncDebugDumpWithNoStack,
+          std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
+          "Flight recorder dump in heartbeatMonitor",
+          false,
+          true);
+    }
```
nit: maybe consider sharing code between the true and false cases and put them in a `do-while` loop?

c-p-i-o replied:
Good suggestion:
Used the following to reuse code.
```
    bool captureStackTrace = true;
    for (int i = 0; i < 2; i++) {
      std::future<bool> asyncDebugDump = std::async(
          std::launch::async, [this, captureStackTrace]() { return this->dumpDebuggingInfo(captureStackTrace); });

      // wait for the dump until timeout - log data
      auto result = waitForFutureOrTimeout(
          asyncDebugDump,
          std::chrono::milliseconds(waitTimeoutDumpInMilSec_),
          "Flight recorder dump in heartbeatMonitor",
          false,
          true);

      if (result) {
        break;        
      }
      // If we failed to dump, try dumping without stack trace.
      captureStackTrace = false;
    }
```

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2007,11 +2027,14 @@ void ProcessGroupNCCL::broadcastDumpSignal() {
+    // stack traces.
     auto status = promiseFlightRecorderDump_.get_future().wait_for(
-        std::chrono::milliseconds(waitTimeoutDumpInMilSec_));
+        std::chrono::milliseconds(2 * waitTimeoutDumpInMilSec_));
```
nit: `waitTimeoutDumpInMilSec_` is not used elsewhere, maybe we can directly bump its value by 2x? (easier maintenance I guess)

c-p-i-o replied:
So the watchdog needs to give 2 mins for heartbeat thread to try to dump (1 min for with stack trace and if this fails, 1 min without stack trace).

kwen2501 reviewed (approved):
LGTM. Have some minor comments.

c-p-i-o commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1513,18 +1516,21 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 }
 
-bool ProcessGroupNCCL::dumpDebuggingInfo() {
+bool ProcessGroupNCCL::dumpDebuggingInfo(bool captureStackTrace) {
```
Done!

c-p-i-o commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1265,13 +1265,14 @@ void ProcessGroupNCCL::enableCollectivesTiming() {
     bool throwException,
     bool log) {
   std::string errorMsg;
+  bool result = false;
```
Done!

c-p-i-o commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1339,6 +1341,7 @@ void ProcessGroupNCCL::waitForFutureOrTimeout(
   if (throwException && !errorMsg.empty()) {
     C10_THROW_ERROR(DistBackendError, errorMsg);
   }
+  return result;
```
It returns `std::timeout` if it fails to complete. When it throw an exception, it returns `std::exception`

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1513,18 +1516,21 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 }
 
-bool ProcessGroupNCCL::dumpDebuggingInfo() {
+bool ProcessGroupNCCL::dumpDebuggingInfo(bool captureStackTrace/*=true*/) {
```
nit: captureCppStackTrace?

c-p-i-o replied:
renamed to `includeStackTrace` to be more clear. Thanks for the suggestion!

fduwjj reviewed (commented):
Do you want to do a roll out on this or it is ok that we directly have it in the code?

c-p-i-o committed:
[RFC][BE] Improve Flight Recorder efficacy (#142178)

Summary:

This is an attempt to improve the flight recorder efficacy. 
We have a small subset of jobs that are timing out (i.e. failing to write out FR logs in 1 minute) and some that are throwing a `std::exception - broken promise`.

There are two changes in here.
1. We attempt to write out FR buffer with stack traces. If this fails, we attempt to capture FR buffer again - but this time without stack traces. The assumption here is that FR could be locking up when unwinding stack.
Note, to keep things simple, I'm re-using the same file name for both with/without stack_trace.
2.  Add additional catch statements in the Manifold writer. There might be something going on in here - so we'll get a log statement if this is failing.


TODO:
- there's nothing differentiating in the output that says whether stack traces were omitted purposefully or not.
This info might be useful for the analyzer - so I'll add this in a follow on diff.

Test Plan: Unit tests.

Reviewed By: kwen2501

Differential Revision: D66843194
